### PR TITLE
add pybind11 wrappers for k2.

### PR DESCRIPTION
**Pull request by csukuangfj**
I am not sure whether the directory structure is suitable for K2.

Any suggestions are welcome.

### CODE REVIEW

**Comment by csukuangfj** [comment]
I'll fix the style check later.

**Comment by csukuangfj** [comment]
Add `clang-tidy`.

It turns out `clang-tidy` is more powerful than `cppcheck`. I've fixed many warnings
issued by `clang-tidy`.

**Comment by csukuangfj** [comment]
Ready to merge.

**Comment by danpovey** [comment]
Oops, sorry, there are now conflicts.

File: k2/python/csrc/fsa.cc [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+           [](const Arc &self, const Arc &other) { return self == other; })
+      .def("__lt__",
+           [](const Arc &self, const Arc &other) { return self < other; });
+
+  py::class_<Fsa>(m, "Fsa")
+      .def(py::init<>())
+      .def("num_states", &Fsa::NumStates)
+      .def("__str__", [](const Fsa &self) { return FsaToString(self); })
+      .def_readwrite("arc_indexes", &Fsa::arc_indexes)
+      .def_readwrite("arcs", &Fsa::arcs);
```
Comment (danpovey) [inline k2/python/csrc/fsa.cc:31]: FYI, the objects we'll mostly be dealing with are the FsaVec, which hasn't been fully fleshed out yet, but this is a great start.

**Comment by qindazhu** [comment]
> Add `clang-tidy`.
> 
> It turns out `clang-tidy` is more powerful than `cppcheck`. I've fixed many warnings
> issued by `clang-tidy`.

There are conflicts. And maybe  you can remove `cppcheck`.

**Comment by csukuangfj** [comment]
Wil resolve it tonight.

**Comment by csukuangfj** [comment]
Let's keep cppcheck right now. I will build a docker image with all dependencies preinstalled.

**Comment by csukuangfj** [comment]
@danpovey @qindazhu 

please have a look at the diagnostic messages issued by clang-tidy
which cause the style check to fail.

I think the messages are very helpful.

File: k2/python/csrc/fsa.cc (hunk 1 above)
Comment (csukuangfj) [reply k2/python/csrc/fsa.cc:31]: Conflicts resolved.

CI fails because of `determinize.cc`. Please have a look at the error messages.

**Comment by danpovey** [comment]
I'm OK with style issues for now.  I will finish the determinize code and
then worry.

**Comment by danpovey** [comment]
I'm OK with style issues for now.  I will finish the determinize code and then worry.
[reposting from github as I dont trust the email interface]

**Comment by qindazhu** [comment]
Seems it has conflict now.

**Comment by csukuangfj** [comment]
resolved.

**Comment by qindazhu** [comment]
Merging, as we will ignore style issues in `determinize.cc` for now.

**Pull request merged**